broker-core/src/main/java/io/zeebe/broker/subscription/message/processor/OpenMessageSubscriptionProcessor.java

```diff
@@ -73,33 +76,46 @@ public void processRecord(
       return;
     }
 
-    // handle new subscription
-    handleNewSubscription(record, streamWriter, sideEffect, subscription);
+    handleNewSubscription(record, streamWriter);
   }
 
   private void handleNewSubscription(
-      TypedRecord<MessageSubscriptionRecord> record,
-      TypedStreamWriter streamWriter,
-      Consumer<SideEffectProducer> sideEffect,
-      MessageSubscription subscription) {
-    final Message message =
-        messageStateController.findFirstMessage(
-            subscriptionRecord.getMessageName(), subscriptionRecord.getCorrelationKey());
+      TypedRecord<MessageSubscriptionRecord> record, TypedStreamWriter streamWriter) {
+
+    sideEffect.accept(this::sendAcknowledgeCommand);
+
+    messageStateController.findMessage(
+        subscriptionRecord.getMessageName(),
+        subscriptionRecord.getCorrelationKey(),
+        this::correlateMessage);
 
-    if (message != null) {
+    messageStateController.put(subscription);
+
+    streamWriter.writeFollowUpEvent(
+        record.getKey(), MessageSubscriptionIntent.OPENED, subscriptionRecord);
+  }
+
+  private boolean correlateMessage(Message message) {
+    // correlate the first message which matches
+    // - ignore messages which are correlated to the workflow instance before
+
+    final boolean isCorrelatedBefore =
```

Comment:
To recap: this means you check if the message was correlated before to this wf instance, since each message has an unique key. This also means that a message with the same correlation key and name can be correlated again against this wf instance. Is this what you expect/want?

Reply:
Correct.